models/autotetraploid.slim

```diff
@@ -0,0 +1,130 @@
+
+*/
+
+initialize() {
+	initializeSLiMModelType("nonWF");
+		
+	defineConstant("K", 100);
+	defineConstant("L", 1e6); // total bp length
+
+	
+	initializeMutationType("m1", 0.5, "e", 0.01);
+	m1.convertToSubstitution = T;
+	
+	initializeGenomicElementType("g1", m1, 1.0);
+	initializeGenomicElement(g1, 0, L - 1);
+	initializeMutationRate(1e-7);
+	initializeRecombinationRate(1e-8);
+	defineConstant("totalLength", L * 1e-8); // total genetic length
+}
+
+reproduction(p1) {
+	mate = p1.sampleIndividuals(1);
+	ma_chroms = c(individual.genomes, p2.individuals[individual.tag].genomes);
+	pa_chroms = c(mate.genomes, p2.individuals[mate.tag].genomes);
+	pairing = sample(seqLen(4), 4);
+	// including 0.0 starts with the second chromosome instead of the first
+	breakpoints1 = asInteger(c(runif(1) < 0.5 ? 0 else NULL, floor(L * runif(rpois(1, totalLength)))));
+	breakpoints2 = asInteger(c(runif(1) < 0.5 ? 0 else NULL, floor(L * runif(rpois(1, totalLength)))));
+	offspring = p1.addRecombinant(ma_chroms[0], pa_chroms[pairing[0]], breakpoints1,
+											ma_chroms[1], pa_chroms[pairing[1]], breakpoints2);
+	breakpoints1 = asInteger(c(runif(1) < 0.5 ? 0 else NULL, floor(L * runif(rpois(1, totalLength)))));
+	breakpoints2 = asInteger(c(runif(1) < 0.5 ? 0 else NULL, floor(L * runif(rpois(1, totalLength)))));
+	other = p2.addRecombinant(ma_chroms[2], pa_chroms[pairing[2]], breakpoints1,
+											ma_chroms[3], pa_chroms[pairing[3]], breakpoints2);
+	tag = runif(1);
+	offspring.tagF = tag;
+	other.tagF = tag;
+}
+
+mutation(m1) {
+	// the fitness effect of having k alleles will be h[k] * s
+	mut.setValue("h", c(0.0, sort(runif(3)), 1.0));
+	return T;
+}
+
+1 early() {
+	sim.addSubpop("p1", K);
+	sim.addSubpop("p2", p1.individualCount);
+	for (k in seqLen(p1.individualCount)) {
+		// we can't point to the other individual so we have to associate them with unique tags
+		tag = runif(1);
+		p1.individuals[k].tagF = tag;
+		p2.individuals[k].tagF = tag;
+		p1.individuals[k].tag = k;
+	}
+}
+
+early() {
+	// initialize tags for new inds
+	inds = p1.individuals[p1.individuals.age == 0];
+	others = p2.individuals[p2.individuals.age == 0];
+	idx = match(inds.tagF, others.tagF);
+	inds.tag = others.index[idx];
+}
+
+2: early() {
+	// mark no-longer-alive p2 individuals for death
+	// don't have to do this every generation
+	alive = rep(0.0, p2.individualCount);
+	alive[p1.individuals.tag] = 1.0;
+	p2.individuals.fitnessScaling = alive;
+	// update tags to reflect this
+	// note we are assuming that death won't rearrange their relative orders
+	tag_map = asInteger(cumSum(alive) - 1);
+	p1.individuals.tag = tag_map[p1.individuals.tag];
+}
+
+early() {
+	p1.fitnessScaling = K / p1.individualCount;
+}
+
+fitness(m1, p1) {
+	other = p2.individuals[individual.tag];
+	h = mut.getValue("h");
+	n = sum(individual.genomes.containsMutations(mut)) + sum(other.genomes.containsMutations(mut));
```

Comment:
Instead of sum(individual.genomes.containsMutations(mut)) you could leverage the heterozygous flag, which would be faster than searching the genomes for the mutation.

Reply:
Hm: what do you mean?

Reply:
Rather than doing sum(individual.genomes.containsMutations(mut)), which searches the two genomes of individual for mut, you could replace that expression with (heterozygous ? 1 else 2), which should be equivalent.  Of course the search through other's genomes is unavoidable.

Reply:
Hm, but what if the individual is homozygous and lacking this mutation? then the callback is not called? Ah, I see.

For clarify, I think I'll keep it this way, but make a note.

Reply:
Yes, the callback would not then be called.  I think I would write it the "right" (i.e., efficient) way and make a note in the opposite direction, but as you wish.  :->

Reply:
let's leave it.